presto-tablestore/src/main/java/com/facebook/presto/tablestore/TablestoreFacade.java

```diff
@@ -0,0 +1,2003 @@
+import static com.facebook.presto.common.predicate.Marker.Bound.BELOW;
+import static com.facebook.presto.common.predicate.Marker.Bound.EXACTLY;
+import static com.facebook.presto.spi.StandardErrorCode.GENERIC_INTERNAL_ERROR;
+import static com.facebook.presto.tablestore.IndexFirstSet.Type.threshold;
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkNotNull;
+import static com.google.common.base.Preconditions.checkState;
+import static io.airlift.slice.Slices.utf8Slice;
+import static io.airlift.slice.Slices.wrappedBuffer;
+import static java.lang.String.format;
+import static java.lang.System.currentTimeMillis;
+import static java.lang.System.nanoTime;
+import static java.util.Collections.emptyList;
+import static java.util.Collections.emptyMap;
+import static java.util.stream.Collectors.toList;
+
+@SuppressWarnings("OptionalUsedAsFieldOrParameterType")
+public class TablestoreFacade
+        extends TablestoreSessionProperties
+        implements AutoCloseable
+{
+    private static final Logger log = Logger.get(TablestoreFacade.class);
+    public static final int MAX_RANGE_TO_POINT_SIZE = 100;
+    private final SyncClient syncClient;
+    private final TablestoreConfig config;
+
+    public TablestoreFacade(TablestoreConfig config)
+    {
+        this(config, TablestoreClientHelper.createSyncClient(config, Optional.empty()));
+        log.info("create a new SyncClient, connectorId=tablestore, config=%s", config);
+    }
+
+    public TablestoreFacade(TablestoreConfig config, SyncClient syncClient)
+    {
+        this.config = config;
+        this.syncClient = syncClient;
+    }
+
+    public <R> Triple<R, Long, Long> executeRemote(Function<SyncClient, R> call,
+            String method, Object... params)
+    {
+        Exception cause;
+        String type;
+        boolean invalidate = false;
+        String traceId = null;
+        String requestId = null;
+        try {
+            long t1 = nanoTime();
+            long t2 = nanoTime();
+            R r = call.apply(syncClient);
+            long t3 = nanoTime();
+            log.info(
+                    "Getting a SyncClient from cache and remote call %s(), clientCosts=%sns apiCost=%sns.",
+                    method, (t2 - t1), (t3 - t2));
+            return Triple.of(r, (t2 - t1), t3 - t2);
+        }
+        catch (TableStoreException e) {
+            type = "table_store_error";
+            traceId = e.getTraceId();
+            requestId = e.getRequestId();
+            cause = e;
+        }
+        catch (ClientException e) {
+            type = "network_abnormal_error";
+            traceId = e.getTraceId();
+            cause = e;
+        }
+        catch (Exception e) {
+            type = "others_error";
+            cause = e;
+            invalidate = true; //有可能客户端被关闭之类的情况
```

Comment:
What happens if the client has been shutdown?  
What's 'invalidate' for? I couldn't find any usage in this class.

Reply:
Indeed not needed, will delete it.